dali/pipeline/operators/arithmetic/expression_impl_gpu.cuh

```diff
@@ -0,0 +1,118 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_ARITHMETIC_EXPRESSION_IMPL_GPU_CUH_
+#define DALI_PIPELINE_OPERATORS_ARITHMETIC_EXPRESSION_IMPL_GPU_CUH_
+
+#include <vector>
+
+#include "dali/core/any.h"
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/operators/arithmetic/arithmetic_meta.h"
+#include "dali/pipeline/operators/arithmetic/expression_impl_factory.h"
+#include "dali/pipeline/operators/arithmetic/expression_tree.h"
+#include "dali/pipeline/operators/op_spec.h"
+#include "dali/pipeline/util/backend2workspace_map.h"
+#include "dali/pipeline/workspace/workspace.h"
+
+namespace dali {
+template <bool IsTensor, typename T>
+using in_desc_t = std::conditional_t<IsTensor, const T*, T>;
+
+// Using the distinction between the Tensors and Constant
+// Otherwise void* could be used.
+template <typename Result, typename Left, bool LeftTensor, typename Right, bool RightTensor>
+struct GPUTileDesc {
+  Result *result;
+  in_desc_t<LeftTensor, Left> left;
+  in_desc_t<RightTensor, Right> right;
+  int64_t extent;
+};
+
+template <ArithmeticOp op, typename Result, typename Left, typename Right>
+__device__ void ExecuteBin(Result *result, const Left *l, const Right *r, int64_t extent) {
+  using meta = arithm_meta<op, GPUBackend>;
+  for (int64_t i = blockIdx.x * blockDim.x + threadIdx.x; i < extent; i += blockDim.x * gridDim.x) {
+    result[i] = meta::impl(l[i], r[i]);
+  }
+}
+
+template <ArithmeticOp op, typename Result, typename Left, typename Right>
+__device__ void ExecuteBin(Result *result, Left l, const Right *r, int64_t extent) {
+  using meta = arithm_meta<op, GPUBackend>;
+  for (int64_t i = blockIdx.x * blockDim.x + threadIdx.x; i < extent; i += blockDim.x * gridDim.x) {
+    result[i] = meta::impl(l, r[i]);
+  }
+}
+
+template <ArithmeticOp op, typename Result, typename Left, typename Right>
+__device__ void ExecuteBin(Result *result, const Left *l, Right r, int64_t extent) {
+  using meta = arithm_meta<op, GPUBackend>;
+  for (int64_t i = blockIdx.x * blockDim.x + threadIdx.x; i < extent; i += blockDim.x * gridDim.x) {
+    result[i] = meta::impl(l[i], r);
+  }
+}
+
+template <ArithmeticOp op, typename Tile>
+__global__ void ExecuteTiled(const Tile *tiles, int num_tiles) {
+  const auto &tile = tiles[blockIdx.y];
+  ExecuteBin<op>(tile.result, tile.left, tile.right, tile.extent);
+}
+
+dim3 GetBlockLayout(int extent, int thread_num, int tiles) {
```

Comment:
```suggestion
dim3 GetGridLayout(int extent, int thread_num, int tiles) {
```
or
```suggestion
dim3 GetGridDim(int extent, int thread_num, int tiles) {
```

Reply:
...also: why do you need thread_num here?

Reply:
thread_num is leftover. Initially I was selecting the grid dim based on number of threads.